sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/HL7v2IO.java

```diff
@@ -0,0 +1,597 @@
+            context.output(HL7v2Message.fromModel(fetchMessage(this.client, msgId)));
+          } catch (Exception e) {
+            failedMessageGets.inc();
+            LOG.warn(
+                String.format(
+                    "Error fetching HL7v2 message with ID %s writing to Dead Letter "
+                        + "Queue. Cause: %s Stack Trace: %s",
+                    msgId, e.getMessage(), Throwables.getStackTraceAsString(e)));
+            context.output(HL7v2IO.Read.DEAD_LETTER, HealthcareIOError.of(msgId, e));
+          }
+        }
+
+        private Message fetchMessage(HealthcareApiClient client, String msgId)
+            throws IOException, ParseException, IllegalArgumentException, InterruptedException {
+          long startTime = System.currentTimeMillis();
+
+          com.google.api.services.healthcare.v1beta1.model.Message msg =
+              client.getHL7v2Message(msgId);
+
+          if (msg == null) {
+            throw new IOException(String.format("GET request for %s returned null", msgId));
+          }
+          this.successfulHL7v2MessageGets.inc();
+          return msg;
+        }
+      }
+    }
+  }
+
+  /** List HL7v2 messages. */
+  public static class ListHL7v2Messages extends PTransform<PBegin, PCollection<HL7v2Message>> {
+    private final List<String> hl7v2Stores;
+    private final String filter;
+
+    /**
+     * Instantiates a new List HL7v2 message IDs with filter.
+     *
+     * @param hl7v2Stores the HL7v2 stores
+     * @param filter the filter
+     */
+    ListHL7v2Messages(List<String> hl7v2Stores, String filter) {
+      this.hl7v2Stores = hl7v2Stores;
+      this.filter = filter;
+    }
+
+    ListHL7v2Messages(List<String> hl7v2Stores) {
+      this.hl7v2Stores = hl7v2Stores;
+      this.filter = null;
+    }
+
+    @Override
+    public PCollection<HL7v2Message> expand(PBegin input) {
+      return input
+          .apply(Create.of(this.hl7v2Stores))
+          .apply(ParDo.of(new ListHL7v2MessagesFn(this.filter)));
+    }
+  }
+
+  static class ListHL7v2MessagesFn extends DoFn<String, HL7v2Message> {
+
+    private final String filter;
+    private transient HealthcareApiClient client;
+
+    /**
+     * Instantiates a new List HL7v2 fn.
+     *
+     * @param filter the filter
+     */
+    ListHL7v2MessagesFn(String filter) {
+      this.filter = filter;
+    }
+
+    /**
+     * Init client.
+     *
+     * @throws IOException the io exception
+     */
+    @Setup
+    public void initClient() throws IOException {
+      this.client = new HttpHealthcareApiClient();
+    }
+
+    /**
+     * List messages.
+     *
+     * @param context the context
+     * @throws IOException the io exception
+     */
+    @ProcessElement
+    public void listMessages(ProcessContext context) throws IOException {
+      String hl7v2Store = context.element();
+      // Output all elements of all pages.
+      this.client.getHL7v2MessageStream(hl7v2Store, this.filter).forEach(context::output);
```

Comment:
Yeah, this can be a performance bottleneck and this whole operation will be limited to a single machine. Usually sources need an additional level of parallelism due to being high fanout. BTW it might sense to add a Reshuffle at the end here just to allow any subsequent steps to parallelize.